doc/good-practices.mld

```diff
@@ -0,0 +1,405 @@
+
+- First, create a quoter using the {{!Ppxlib.Expansion_helpers.Quoter.create}[create]} function:
+
+{[
+  # open Expansion_helper ;;
+  #s let quoter = Quoter.create () ;;
+  val quoter : Quoter.t = <abstr>
+]}
+
+- Then, use {{!Ppxlib.Expansion_helpers.Quoter.quote}[Expansion_helpers.Quoter.quote]} to quote all the expressions that are given from the user, might rely on a context, and that you want "intact."
+
+{[
+  # let quoted_part = Quoter.quote quoter part_to_quote ;;
+  val quoted_payload : expression =
+]}
+
+- Finally, call {{!Ppxlib.Expansion_helpers.Quoter.sanitize}[Expansion_helpers.Quoter.sanitize]} on the whole expression (with quoted parts).
+
+{[
+  # let result = Expansion_helpers.Quoter.sanitize ~quoter rewritten_expression ;;
+  val result : expression =
+  ...
+]}
+
+If the [debug] rewriter had been written using this method, the quoting would
+have ensured that the payload is evaluated in the same context as the
+extension node!
+
+Here is an example on how to write a [debug] rewriter (with the limitation that the payload should not contain variable binding, but the code was left simple to illustrate quoting):
+
+{[
+# let rewrite expr =
+    (* Create a quoter *)
+    let quoter = Quoter.create () in
+    (* An AST mapper to log and replace variables with quoted ones *)
+    let replace_var =
+      object
+        (* See the chapter on AST traverse *)
+        inherit Ast_traverse.map as super
+  
+        (* in case of expression *)
+        method! expression expr =
+          match expr.pexp_desc with
+          (* in case of identifier (not "+") *)
+          | Pexp_ident { txt = Lident var_name; loc }
+            when not (String.equal "+" var_name) ->
+              (* quote the var *)
+              let quoted_var = Quoter.quote quoter expr in
+              let name = Ast_builder.Default.estring ~loc var_name in
+              (* and rewrite the expression *)
+              [%expr
+                debug [%e name] [%e quoted_var];
+                [%e quoted_var]]
+          (* otherwise, continue inside recursively *)
+          | _ -> super#expression expr
+      end
+    in
+    let quoted_rewrite = replace_var#expression expr in
+    let loc = expr.pexp_loc in
+    (* Sanitize the whole thing *)
+    Quoter.sanitize quoter
+      [%expr
+        let debug = Printf.printf "%s = %d; " in
+        [%e quoted_rewrite]] ;;
+  val rewrite : expression -> expression = <fun>
+]}
+
+With {!Ppxlib}'s current quoting mechanism, the code given in that example would look like:
+
+{[
+  # Format.printf "%a\n" Pprintast.expression @@ rewrite [%expr debug + 1, y + 2] ;;
+  let rec __1 = y
+  and __0 = debug in
+  let debug = Printf.printf "%s = %d; " in
+  (((debug "debug" __0; __0) + 1), ((debug "y" __1; __1) + 2))
+  - : unit = ()
+]}
+
+{1:resp_loc Respecting Locations}
+
+Correctly dealing with location is essential to correctly generate OCaml code.
```

Comment:
Yes, maybe we should reorganize the order to give more importance to dealing with locations!

However, I don't think we should just say "avoid `Location.none`", but rather "`Location.none` serves this purpose and should only be used in those cases".
For instance, maybe it is better to use `Location.none` rather than having an unrelated location? So that, if the user has an error, at least it knows the location is not relevant?
I would be happy to have some hints on this!

Reply:
That's a good point. The only case I can think of where it makes sense to add `Location.none` is when pprinting the generated AST (so that it will get re-parsed by the OCaml compiler with the compiler locations). Btw, that's another section we could add to the manual (for now also as a "not yet written" section. What do you think?

Reply:
After thinking a bit about it, you might be right that there is no situation where it is actually useful: it is only useful when building nodes completely from scratch.
This does happen sometimes in the compiler, but never in ppxlib: at least, we have a file name, and very often more.
So it makes sense to explain why always avoiding Location.none in the ppxlib context.

I am not aware of this mechanism (pprinting-reparsing). I think it could be added as a "to-be-written" section, but hopefully with a small explanation for it not to be too mysterious! (It is a bit mysterious right now for me ;))